drivers/xen/regions.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/xen/generic.h>
+
+#include <kernel_arch_interface.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/arch_interface.h>
+#include <zephyr/sys/mem_blocks.h>
+
+LOG_MODULE_REGISTER(xen_regions);
+
+#define EXTENDED_REGIONS_IDX 1
+#define XEN_HYP_NODE DT_INST(0, xen_xen)
+
+static K_MUTEX_DEFINE(ext_regions_lock);
+
+struct extended_regions {
+	sys_mem_blocks_t *block;
+	uint64_t phys_addr;
+	uint64_t size;
+};
+
+/*
+ * At least one extended region should be provided.
+ * If there is no extended regions - alternative allocation should be implemented
+ * by the driver user and driver should be disabled.
+ */
+BUILD_ASSERT(DT_NUM_REGS(XEN_HYP_NODE) >= EXTENDED_REGIONS_IDX,
+	     "No Extended regions were provided in DT");
+
+/*
+ * Add BUILD_ASSERT that rejects extended region, starting from 0x0.
+ * This is done because mem_blocks is not designed to handle buffer
+ * starting from 0 and has assert with check. In this case starting address
+ * should be shifted to page 1. That shouldn't be a problem because
+ * minimal extended region size is set to 64MB in XEN.
+ * For example for the extended region 0->0x30000000 the following
+ * regs should be provided:
+ * regs = <.... 0x0 0x1000 0x0 0x2FFFF000 ... >;
+ *
+ * Check reg addr and reg size to be aligned to
+ * CONFIG_MMU_PAGE_SIZE. This is the requirement for arch_mem_map call.
+ *
+ * Extended regions should not overlap kernel Virtual Memory region.
+ */
+#define EXT_BLOCK(node_id, idx)						\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),		\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	BUILD_ASSERT(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region size is not aligned");		\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region address is not aligned");		\
+	BUILD_ASSERT((DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) < CONFIG_KERNEL_VM_BASE || \
+		      DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) >=		\
+		      CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE),	\
```

Comment:
it sounds really funny for me especially from you
because a few days ago you created this commit: https://github.com/xen-troops/zephyr/pull/43/commits/f64569c4cbe74162fe5000d2189e22c225f57b62

Reply:
Hm. Probably. But I've got warning for that assert. So what made you think that owerflow is possible? And why no warning here?

Reply:
Rewritten